### PR TITLE
obsolete_statements - posix regex

### PR DESCRIPTION
**Pull request by jan-jezek**
Ref #2132.

Posix regex becomes obsolete in abap version 7.56, however this versions is not yet available to the public.
Should I add new abap version 7.56 for this?

Regex in string function is not in this as I don't know how to easily check built-in methods (and I don't have the full list of methods that use this). Generic parameter name "regex" would create false positives for user defined methods.

### CODE REVIEW

File: packages/core/test/rules/obsolete_statement.ts [hunk 1]
```diff
@@ -74,6 +76,58 @@ END OF li_order.`, cnt: 1, fix: false},
...
+  const config = Config.getDefault(version);
+  const reg = new Registry(config).addFile(new MemoryFile("zfoo.prog.abap", abap));
+  await reg.parseAsync();
+  const rule = new ObsoleteStatement();
+  return rule.initialize(reg).run(reg.getFirstObject()!);
+}
+
+describe("test obsolete_statements rule - versions", () => {
+  it("statements no issues", async () => {
+    const issue1 = await findIssues("FIND REGEX 'foo' IN 'bar'.", Version.v754);
```
Comment (jan-jezek) [inline packages/core/test/rules/obsolete_statement.ts:89]: I don't know how to run the check using different version, other than this.
I remember there being a way to do it with testRule, but I don't know how.

File: packages/core/src/rules/obsolete_statement.ts [hunk 2]
```diff
@@ -287,6 +291,39 @@ CALL TRANSFORMATION OBJECTS: https://help.sap.com/doc/abapdocu_752_index_htm/7.5
...
+            issues.push(issue);
+          }
+        }
+        else {
+          const classNameExpression = staNode.findAllExpressions(Expressions.ClassName);
+          const methodNameExpression = staNode.findAllExpressions(Expressions.MethodName);
+
+          if (classNameExpression.length !== 0 && methodNameExpression.length !== 0) {
+            const className = classNameExpression[0].concatTokens();
+            const methodName = methodNameExpression[0].concatTokens();
```
Comment (jan-jezek) [inline packages/core/src/rules/obsolete_statement.ts:310]: Probably slow, but I don't have better idea.
Checking for 0 is fine as it doesn't make much sense to chain these methods.

**Comment by larshp** [comment]
yes, add 756 in https://github.com/abaplint/abaplint/blob/main/packages/core/src/version.ts

File: packages/core/test/rules/obsolete_statement.ts (hunk 1 above)
Comment (larshp) [reply packages/core/test/rules/obsolete_statement.ts:89]: this should work okay for this scenario?

File: packages/core/src/rules/obsolete_statement.ts (hunk 2 above)
Comment (larshp) [reply packages/core/src/rules/obsolete_statement.ts:310]: handling deprecated parameters for methods should ideally be handled via abapdoc, but I dont think it has annotations for deprecation?

**Comment by larshp** [comment]
built-in: the syntax checker in abaplint builds a [spaghetti tree](https://en.wikipedia.org/wiki/Parent_pointer_tree) and keeps track of references, take a look at https://github.com/abaplint/abaplint/blob/f5e326823eee124c5dddbf80b535e50d53086df7/packages/core/src/rules/downport.ts#L783 

also consider if you'd like to do this as a second step, ie ship these changes first.

let me know if you like more hints 😄

File: packages/core/test/rules/obsolete_statement.ts (hunk 1 above)
Comment (jan-jezek) [reply packages/core/test/rules/obsolete_statement.ts]: This works fine.

File: packages/core/src/rules/obsolete_statement.ts (hunk 2 above)
Comment (jan-jezek) [reply packages/core/src/rules/obsolete_statement.ts:310]: No, unfortunately no annotations for deprecation.
Maybe it should be moved to new rule "obsolete_method" or "obsolete_method_parameter"?

**Comment by larshp** [comment]
regression test failing: need the fixes from https://github.com/abaplint/abaplint/pull/2140, I'll publish that, sec

**Comment by jan-jezek** [comment]
> built-in: the syntax checker in abaplint builds a [spaghetti tree](https://en.wikipedia.org/wiki/Parent_pointer_tree) and keeps track of references, take a look at
> 
> https://github.com/abaplint/abaplint/blob/f5e326823eee124c5dddbf80b535e50d53086df7/packages/core/src/rules/downport.ts#L783
> 
> also consider if you'd like to do this as a second step, ie ship these changes first.
> 
> let me know if you like more hints 😄

I appreciate the hints 😄

**Comment by jan-jezek** [comment]
I will add built-in functions later, no rush as v756 is not even out yet.

**Comment by larshp** [comment]
published in 2.79.0

**Pull request merged**